### ISSUE TITLE
secret_key rotation is failing in Grafana

### ISSUE THREAD

**Issue by alokr** [grafana/grafana#62737]
**What happened**:
We tried rotating the secret_key in Grafana but after rotation we're getting an authorization error. We have tried `grafana-cli admin secrets-migration re-encrypt-data-keys` after the rotation as well but that hasn't helped either. I can see secure_json_data filed in db is getting re-encrypted, but the queries doesn't seem to be using the new key.

**What you expected to happen**:
After secret_key rotation queries should have worked as they were working before rotation.

**How to reproduce it (as minimally and precisely as possible)**:
1. Instantiate Grafana with a secret key.
2. Manually provision a datasource which uses username and password.
3. Test to make sure data source was added successfully.
4. Rotate secret_key.
5. Re-deploy Grafana
6. Queries would start failing with authentication errors.
7. Try re-encrypt-data-keys using Grafana CLI. The queries still fail.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.1.5
- Data source type & version: mysql
- OS Grafana is installed on: Linux
- User OS & Browser: Mac, Chrome
- Grafana plugins: 
- Others:

**Comment by joanlopez** [comment]
Hi @alokr,

I think there has been a misunderstand in the process, therefore room for improving our docs, but let me try to clarify it.

Secrets stored in Grafana DB are encrypted with data encryption keys (DEKs), which are stored as well in the Grafana DB encrypted with a key encryption key (KEK), which could be:
- (a) `secret_key`
- (b) A key configured externally in one of the supported KMS, like AWS KMS (_Enterprise only_).

Therefore, the operation of [re-encrypting data keys](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/#re-encrypt-data-keys) (which consists of decrypting them with the current/old KEK and re-encrypting them with a new KEK -otherwise the operation would be useless) is only supported to transition from (a) to (b) or (b) to (b) (like switching from AWS KMS to Azure Key Vault, or even between different keys within the same KMS), but not (a) to (a), because there's no way to do something like defining two `secret_key`s. You can find more info about that [here](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/#encrypting-your-database-with-a-key-from-a-key-management-service-kms).

That's precisely why you're getting an error, and the reason why that's expected, because if you directly change the current `secret_key` value, encrypted DEKs stored in Grafana DB will no longer be decryptable (step 6 in your description) and therefore you won't be able to re-encrypt them (step 7).

Is that explanation clear enough? Any additional doubts? 

If that's fine, I'll close this issue and recommend you to open a feature request  asking for the support you're looking for (a way to rotate from a manually configured, settings provided `secret_key` to another one).

Thanks!

**Comment by radek-szostak** [comment]
Thank you @joanlopez for you explanation.

Is there a recommended procedure for rotating `secret_key`?

**Comment by joanlopez** [comment]
> Thank you [@joanlopez](https://github.com/joanlopez) for you explanation.
> 
> Is there a recommended procedure for rotating `secret_key`?

Hey @radek-szostak,

I'm not a security expert, but I'd suggest having a look at the operation of re-encrypting data encryption keys -DEKs- (see https://github.com/grafana/grafana/blob/main/pkg/cmd/grafana-cli/commands/secretsmigrations/secretsmigrations.go#L10), as it will basically decrypt the existing DEKs with the former secret key, and re-encrypt them with the new one.

But, please be aware that if you're using a secret key hard-coded in the INI configuration, and you want to update it, you will need to use a KMS -like Vault-, as re-encrypting DEKs between two secret keys manually defined in Grafana config isn't a supported operation, mainly because the use of manually defined secret keys isn't very recomended from a security standpoint.

So, in short, you will likely need to:
1. Define a new encryption key, other than `secret_key`, with a KMS. _For instance, [here](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/encrypt-secrets-using-hashicorp-key-vault/) you can find the instructions specific to Hashicorp Vault, but other systems (like AWS KMS, or Google Cloud KMS are also supported)._
2. Run the CLI command to rotate data encryption keys (DEKs).

I hope that helps! 🙇🏻
